File: core/domain/usage.go
package domain

import "time"

// UsageGrain identifies the aggregation level of a usage rollup row.
type UsageGrain string

const (
	UsageGrainTurn    UsageGrain = "turn"
	UsageGrainSession UsageGrain = "session"
	UsageGrainProject UsageGrain = "project"
	UsageGrainModel   UsageGrain = "model"
	UsageGrainAgent   UsageGrain = "agent"
)

// UsageDelta is a single LLM call's token contribution to rollups.
type UsageDelta struct {
	PromptTokens     int
	CompletionTokens int
	TotalTokens      int
	Model            string
	AgentID          string
}

// Normalize fills TotalTokens from prompt+completion when the API omitted it.
func (d UsageDelta) Normalize() UsageDelta {
	if d.TotalTokens <= 0 {
		d.TotalTokens = d.PromptTokens + d.CompletionTokens
	}
	return d
}

// UsageRollup is one accumulated row (turn / session / project / model / agent).
type UsageRollup struct {
	Grain            UsageGrain `json:"grain"`
	RefID            string     `json:"refId"`
	ProjectID        string     `json:"projectId,omitempty"`
	SessionID        string     `json:"sessionId,omitempty"`
	Model            string     `json:"model,omitempty"`
	AgentID          string     `json:"agentId,omitempty"`
	PromptTokens     int        `json:"promptTokens"`
	CompletionTokens int        `json:"completionTokens"`
	TotalTokens      int        `json:"totalTokens"`
	CallCount        int        `json:"callCount"`
	MaxPromptTokens  int        `json:"maxPromptTokens,omitempty"`
	UpdatedAt        time.Time  `json:"updatedAt"`
}

// UsageSummary is aggregated token totals.
type UsageSummary struct {
	PromptTokens     int `json:"promptTokens"`
	CompletionTokens int `json:"completionTokens"`
	TotalTokens      int `json:"totalTokens"`
	CallCount        int `json:"callCount"`
	MaxPromptTokens  int `json:"maxPromptTokens,omitempty"`
	TurnCount        int `json:"turnCount"`
	AvgTurnTokens    int `json:"avgTurnTokens,omitempty"` // TotalTokens / TurnCount when TurnCount > 0
}

// FinalizeAvgTurnTokens fills AvgTurnTokens from TotalTokens / TurnCount.
func (s *UsageSummary) FinalizeAvgTurnTokens() {
	if s.TurnCount > 0 {
		s.AvgTurnTokens = s.TotalTokens / s.TurnCount
	} else {
		s.AvgTurnTokens = 0
	}
}

// UsageBreakdown is a parent summary plus child rollups.
type UsageBreakdown struct {
	Summary  UsageSummary  `json:"summary"`
	Turns    []UsageRollup `json:"turns,omitempty"`
	Sessions []UsageRollup `json:"sessions,omitempty"`
	Models   []UsageRollup `json:"models,omitempty"`
	Agents   []UsageRollup `json:"agents,omitempty"`
}

// UsagePeriod is the chart bucket size.
type UsagePeriod string

const (
	UsagePeriodDay   UsagePeriod = "day"
	UsagePeriodWeek  UsagePeriod = "week"
	UsagePeriodMonth UsagePeriod = "month"
)

// UsageSeriesFilter selects chart data. Buckets use rollup UpdatedAt.
type UsageSeriesFilter struct {
	Period    UsagePeriod
	ProjectID string
	Model     string
	AgentID   string
	From      time.Time
	To        time.Time
	// Grain defaults to session (sessions SUM'd into period buckets).
	// Use UsageGrainModel to chart per-model rows (filter.Model optional).
	Grain UsageGrain
}

// UsageSeriesPoint is one period bucket.
type UsageSeriesPoint struct {
	PeriodStart      time.Time `json:"periodStart"`
	PromptTokens     int       `json:"promptTokens"`
	CompletionTokens int       `json:"completionTokens"`
	TotalTokens      int       `json:"totalTokens"`
	CallCount        int       `json:"callCount"`
	Model            string    `json:"model,omitempty"`
	AgentID          string    `json:"agentId,omitempty"`
}

// ModelRollupRefID builds the rollup ref for a model row (project-scoped when projectID set).
func ModelRollupRefID(projectID, model string) string {
	if model == "" {
		return ""
	}
	if projectID == "" {
		return model
	}
	return projectID + "\x1f" + model
}

// AgentRollupRefID builds the rollup ref for an agent row (project-scoped when projectID set).
func AgentRollupRefID(projectID, agentID string) string {
	if agentID == "" {
		return ""
	}
	if projectID == "" {
		return agentID
	}
	return projectID + "\x1f" + agentID
}

File: frontend/src/api/usage.ts
import { asArray, fetchJSON } from '@/api/client'

export interface UsageSummary {
  promptTokens: number
  completionTokens: number
  totalTokens: number
  callCount: number
  maxPromptTokens?: number
  turnCount: number
  avgTurnTokens?: number
}

export interface UsageRollup {
  grain: string
  refId: string
  projectId?: string
  sessionId?: string
  model?: string
  agentId?: string
  promptTokens: number
  completionTokens: number
  totalTokens: number
  callCount: number
  maxPromptTokens?: number
  updatedAt: string
}

export interface UsageBreakdown {
  summary: UsageSummary
  turns?: UsageRollup[]
  sessions?: UsageRollup[]
  models?: UsageRollup[]
  agents?: UsageRollup[]
}

export interface UsageSeriesPoint {
  periodStart: string
  promptTokens: number
  completionTokens: number
  totalTokens: number
  callCount: number
  model?: string
  agentId?: string
}

export type UsagePeriod = 'day' | 'week' | 'month'

export async function fetchSessionUsage(sessionId: string): Promise<UsageBreakdown> {
  return fetchJSON<UsageBreakdown>(`/sessions/${sessionId}/usage`)
}

export async function fetchProjectUsage(projectId: string): Promise<UsageBreakdown> {
  return fetchJSON<UsageBreakdown>(`/projects/${projectId}/usage`)
}

export async function fetchUsageSummary(params?: {
  projectId?: string
  model?: string
}): Promise<UsageSummary> {
  const q = new URLSearchParams()
  if (params?.projectId) q.set('project_id', params.projectId)
  if (params?.model) q.set('model', params.model)
  const qs = q.toString()
  return fetchJSON<UsageSummary>(`/usage/summary${qs ? `?${qs}` : ''}`)
}

export async function fetchUsageSeries(params: {
  period?: UsagePeriod
  projectId?: string
  model?: string
  agentId?: string
  grain?: string
  from?: string
  to?: string
}): Promise<UsageSeriesPoint[]> {
  const q = new URLSearchParams()
  if (params.period) q.set('period', params.period)
  if (params.projectId) q.set('project_id', params.projectId)
  if (params.model) q.set('model', params.model)
  if (params.agentId) q.set('agent_id', params.agentId)
  if (params.grain) q.set('grain', params.grain)
  if (params.from) q.set('from', params.from)
  if (params.to) q.set('to', params.to)
  const qs = q.toString()
  const res = await fetchJSON<{ points: UsageSeriesPoint[] | null }>(`/usage/series${qs ? `?${qs}` : ''}`)
  return asArray(res.points)
}

export async function fetchUsageModels(projectId?: string): Promise<UsageRollup[]> {
  const qs = projectId ? `?project_id=${encodeURIComponent(projectId)}` : ''
  const res = await fetchJSON<{ models: UsageRollup[] | null }>(`/usage/models${qs}`)
  return asArray(res.models)
}

export async function fetchUsageAgents(projectId?: string): Promise<UsageRollup[]> {
  const qs = projectId ? `?project_id=${encodeURIComponent(projectId)}` : ''
  const res = await fetchJSON<{ agents: UsageRollup[] | null }>(`/usage/agents${qs}`)
  return asArray(res.agents)
}
